### PR TITLE
Add eventsV2 support for SQL 

### PR DESCRIPTION
**Pull request by longquanzheng**

### CODE REVIEW

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+
+	history := make([]*p.DataBlob, 0, int(request.PageSize))
+	lastNodeID := int64(-1)
+	lastTxnID := int64(-1)
+	eventBlob := &p.DataBlob{}
+
+	for _, row := range rows {
+		eventBlob.Data = row.Data
+		eventBlob.Encoding = common.EncodingType(row.DataEncoding)
+		if row.NodeID < lastNodeID {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: sounds good

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+		// this is the case that new branch's ancestors doesn't include the forking branch
+		for _, br := range forkB.Ancestors {
+			if *br.EndNodeID >= request.ForkNodeID {
+				newAncestors = append(newAncestors, &shared.HistoryBranchRange{
+					BranchID:    br.BranchID,
+					BeginNodeID: br.BeginNodeID,
+					EndNodeID:   common.Int64Ptr(request.ForkNodeID),
+				})
+				break
+			} else {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: Not really, there is a break before else

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+		}
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return err
+		}
+		if rowsAffected != 1 {
+			return fmt.Errorf("expected 1 row to be affected for tree table, got %v", rowsAffected)
+		}
+		return nil
+	} else {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: OK

File: schema/mysql/v57/cadence/schema.sql [hunk 4]
```diff
@@ -371,4 +373,27 @@ failover_version,
...
+	branch_id      BINARY(16) NOT NULL,
+	node_id BIGINT NOT NULL,
+	txn_id BIGINT NOT NULL,
+	data MEDIUMBLOB NOT NULL,
+	data_encoding  VARCHAR(16) NOT NULL,
+	PRIMARY KEY (tree_id, branch_id, node_id, txn_id)
+);
+
+-- history eventsV2: history_tree stores branch metadata
+CREATE TABLE history_tree (
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply schema/mysql/v57/cadence/schema.sql]: OK

File: common/persistence/sql/storage/mysql/eventsV2.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+const (
+	// below are templates for history_node table
+	addHistoryNodesQry = `INSERT INTO history_node (` +
+		`tree_id, branch_id, node_id, txn_id, data, data_encoding) ` +
+		`VALUES (:tree_id, :branch_id, :node_id, :txn_id, :data, :data_encoding) `
+
+	getHistoryNodesQry = `SELECT node_id, txn_id, data, data_encoding FROM history_node ` +
+		`WHERE tree_id = ? AND branch_id = ? AND node_id >= ? and node_id < ? LIMIT ? `
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/storage/mysql/eventsV2.go]: let's talk offline for it

File: schema/mysql/v57/cadence/schema.sql [hunk 6]
```diff
@@ -371,4 +373,27 @@ failover_version,
 is_global_domain,
 active_cluster_name) values(
 UNHEX('32049b68787240948e63d0dd59896a83'),
-'cadence-system', 0, 'cadence system workflow domain', 'cadence-dev-group@uber.com', 3, 0, '', 0, 0, 0, 0, 0, 0, "");
\ No newline at end of file
+'cadence-system', 0, 'cadence system workflow domain', 'cadence-dev-group@uber.com', 3, 0, '', 0, 0, 0, 0, 0, 0, "");
+
+-- history eventsV2: history_node stores history event data
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply schema/mysql/v57/cadence/schema.sql]: OK

File: schema/mysql/v57/cadence/schema.sql [hunk 7]
```diff
@@ -371,4 +373,27 @@ failover_version,
...
+CREATE TABLE history_node (
+	tree_id      BINARY(16) NOT NULL,
+	branch_id      BINARY(16) NOT NULL,
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply schema/mysql/v57/cadence/schema.sql]: OK

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+			rowsAffected, err = result.RowsAffected()
+			if err != nil {
+				return err
+			}
+			if rowsAffected != 1 {
+				return fmt.Errorf("expected 1 row to be affected for tree table, got %v", rowsAffected)
+			}
+			return nil
+		})
+	} else {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: OK

File: common/persistence/sql/storage/mysql/eventsV2.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/storage/mysql/eventsV2.go]: Confirmed that MySQL doesn't guarantee ordering by PK, I will add order by.

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+			BranchID:   sqldb.MustParseUUID(branchInfo.GetBranchID()),
+			InProgress: false,
+			CreatedTs:  time.Now(),
+			Ancestors:  ancestors,
+			Info:       request.Info,
+		}
+
+		return m.txExecute("AppendHistoryNodes", func(tx sqldb.Tx) error {
+			result, err := tx.InsertIntoHistoryNode(nodeRow)
+			if err != nil {
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: Double checked that our application layer guarantees that no duplicated rows will be inserted.

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+			result, err := tx.DeleteFromHistoryNode(nodeFilter)
+			if err != nil {
+				return err
+			}
+			rowsAffected, err := result.RowsAffected()
+			if err != nil {
+				return err
+			}
+			if rowsAffected <= 0 {
+				return fmt.Errorf("expected >=1 row to be affected for node table, got %v", rowsAffected)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: This is a good point. There can be a case the Fork succeeds but append data node fails. So deleting data node might affect 0 rows.

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+		result, err := m.db.UpdateHistoryTree(row)
+		if err != nil {
+			return err
+		}
+		rowsAffected, err := result.RowsAffected()
+		if err != nil {
+			return err
+		}
+		if rowsAffected != 1 {
+			return fmt.Errorf("expected 1 row to be affected for tree table, got %v", rowsAffected)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: confirmed that it shouldn't happen

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+			result, err = tx.DeleteFromHistoryTree(treeFilter)
+			if err != nil {
+				return err
+			}
+			rowsAffected, err = result.RowsAffected()
+			if err != nil {
+				return err
+			}
+			if rowsAffected != 1 {
+				return fmt.Errorf("expected 1 row to be affected for tree table, got %v", rowsAffected)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: confirmed that it shouldn't happen

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+		TreeID:     sqldb.MustParseUUID(treeID),
+		BranchID:   sqldb.MustParseUUID(request.NewBranchID),
+		InProgress: true,
+		CreatedTs:  time.Now(),
+		Ancestors:  ancestors,
+		Info:       request.Info,
+	}
+	result, err := m.db.InsertIntoHistoryTree(row)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: confirmed that it shouldn't happen

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 14]
```diff
@@ -0,0 +1,508 @@
...
+		}
+	}
+
+	return m.txExecute("DeleteHistoryBranch", func(tx sqldb.Tx) error {
+		branchID := sqldb.MustParseUUID(*branch.BranchID)
+		treeFilter := &sqldb.HistoryTreeFilter{
+			TreeID:   sqldb.MustParseUUID(treeID),
+			BranchID: &branchID,
+		}
+		_, err := tx.DeleteFromHistoryTree(treeFilter)
```
[the remark this answers is not in the document]
Comment (longquanzheng) [reply common/persistence/sql/sqlHistoryV2Manager.go]: addressed offline

File: common/persistence/sql/sqlHistoryV2Manager.go [hunk 15]
```diff
@@ -0,0 +1,502 @@
...
+	branchID := sqldb.MustParseUUID(*branch.BranchID)
+	if request.Success {
+		row := &sqldb.HistoryTreeRow{
+			TreeID:     treeID,
+			BranchID:   branchID,
+			InProgress: false,
+		}
+		result, err := m.db.UpdateHistoryTree(row)
+		if err != nil {
+			return err
```
Comment (venkat1109) [inline common/persistence/sql/sqlHistoryV2Manager.go]: wrap this in InternalServiceError - if you want the caller to retry on this error

**Pull request merged**